Pod CPU usage check from Windows node summary

knabben opened the pull request:
#### What type of PR is this?
/kind bug
/sig windows

#### What this PR does / why we need it:
Moving CPU usage and stats summary to a Polling function with 2 minutes delay (this cover the other behavior of time.sleep 2 minutes), if limit exceed happens or CPU usage is 0, 2 more retries will happen before timing out, this can reduce the flakes.

Could not replicate the issue on a vSphere cluster, kubelet stats seems stable. Need to confirm the `containerd` version, here is `1.6.24` to try to replicate again.

#### Which issue(s) this PR fixes:
Fixes #122092

knabben committed:
CPU usage check from node summary endpoint with retry and checked within 3 minutes

knabben commented:
/test

knabben commented:
/test pull-kubernetes-verify

knabben commented:
/test pull-kubernetes-e2e-capz-windows-master

jsturtevant commented:
/triage accepted
/priority important-soon

jsturtevant commented on test/e2e/windows/cpu_limits.go:
```diff
@@ -46,63 +46,61 @@ var _ = sigDescribe(feature.Windows, "Cpu Resources", framework.WithSerial(), sk
+			framework.ExpectNoError(err, "Error waiting for pods entering on running state.")
+
+			ginkgo.By("Gathering node summary stats")
+			err = wait.PollUntilContextTimeout(ctx, time.Minute*2, time.Minute*6, false, func(ctx context.Context) (bool, error) {
```
could we add a note here about this not triggering until 2 mins go passed, then ever 2 mins?

jsturtevant reviewed (commented):
The upstream test is using containerd 1.7. Maybe that is the difference on vSpere?

Seems strange we get 0 after the pods run for 2 mins. Any ideas where it might be failing? Is the get summary failing and so we end up with 0? or is the actual reported value for the pod zero?

knabben commented:
> The upstream test is using containerd 1.7. Maybe that is the difference on vSpere?
> 
> Seems strange we get 0 after the pods run for 2 mins. Any ideas where it might be failing? Is the get summary failing and so we end up with 0? or is the actual reported value for the pod zero?

Yes, I saw this behavior with the pod running in the first <10 seconds, not after after the sleep. The pod was still Running when the test failed from the dump, so I suppose the value being reported on stats summary is failing (could replicate neither, the CPU usage was always ~500m after the sleep), will try again with 1.7.

jsturtevant commented:
> so I suppose the value being reported from stats summary is the failure point (could replicate neither, the CPU usage was always ~500m after the sleep),

if this is the case we might be able to wrap that call in retry instead of waiting an additional 2 mins?

knabben commented:
For this statement is a matter of changing the call to:

```
immediate := true
wait.PollUntilContextTimeout(ctx, 30 * time.Second, 6 * time.Minute, immediate, func()..
```

knabben commented:
> > If we do a retry call (maybe after 30 seconds) at this point, and check the ~500m value it seems a good test IMO
> 
> I generally agree, This test should start up the pod, and once cpu is being consumed we should be verifing that it doesn't go over 500. If that is the case we could so something like gomega.eventually and consistently to achieve that as described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts

Sure changing it for gomega.Eventually usage. 

> > If we do a retry call (maybe after 30 seconds) at this point, and check the ~500m value it seems a good test IMO
> 
> I generally agree, This test should start up the pod, and once cpu is being consumed we should be verifing that it doesn't go over 500. If that is the case we could so something like gomega.eventually and consistently to achieve that as described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts

Makes sense in terms of reliability, changed for both functions usage.

knabben commented:
> > If we do a retry call (maybe after 30 seconds) at this point, and check the ~500m value it seems a good test IMO
> 
> I generally agree, This test should start up the pod, and once cpu is being consumed we should be verifing that it doesn't go over 500. If that is the case we could so something like gomega.eventually and consistently to achieve that as described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts


Makes sense in terms of reliability, changed for both functions usage.
Renamed the function name as well

knabben commented:
> I generally agree, This test should start up the pod, and once cpu is being consumed we should be verifing that it doesn't go over 500. If that is the case we could so something like gomega.eventually and consistently to achieve that as described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts

Makes sense in terms of reliability, changed for both functions usage. Function renamed

jsturtevant commented:
We noticed during sig-windows triage today that we are only see this error on 1.7+. Our 1.6 tests cluster does not have this flake.  @knabben can you see if the changes work against a 1.7 vSphere  cluster or if you can reproduce the error before these changes?

knabben commented:
Running with 1.7.6 on Windows CAPV, these are the errors, seem flaky with the official `e2e.test 1.18.0`:

``` 
[FAILED] Pod cpu-resources-test-windows-8357/cpulimittest-1936531e-6dd0-4d36-9d79-01a9a64c8534 reported usage is 1.06193494, but it should not exceed limit by > 5%
  In [It] at: test/e2e/windows/cpu_limits.go:98 @ 12/12/23 15:41:52.014
```

```
[FAILED] Pod cpu-resources-test-windows-1393/cpulimittest-b6954b14-fc83-4362-89db-737172c07e06 reported usage is 0.6330031110000001, but it should not exceed limit by > 5%
  In [It] at: test/e2e/windows/cpu_limits.go:98 @ 12/12/23 15:49:07.446
```

```
  STEP: Gathering node summary stats @ 12/12/23 16:13:43.427
  Dec 12 16:13:43.719: INFO: Pod cpulimittest-00ece286-5207-4182-90f9-26c3344cbdf0 usage: 0
  [FAILED] in [It] - test/e2e/windows/cpu_limits.go:95 @ 12/12/23 16:13:43.719
```

jsturtevant commented:
those look consistent with what we are seeing in capz (https://github.com/kubernetes/kubernetes/issues/122092). This means it is likely a bug in containerd or the processing of the stats in kubelet

knabben commented:
/hold

knabben commented:
Removing the `Consistently` statement with 1.7 this wont pass, letting the Eventually statement to allow the retry in both 0 and >500ms  cases within 2 minutes.

/unhold

jsturtevant commented:
> Removing the `Consistently` statement with 1.7 this wont pass, letting the Eventually statement to allow the retry in both 0 and >500ms cases within 2 minutes.

It seems like this is hiding a bug somewhere else in the stack, we should track that down instead of adjusting the test to allow for inconsistency.

jsturtevant commented:
We've tracked down the bug to containerd: https://github.com/containerd/containerd/issues/9531
